imap/mailbox.c

```diff
@@ -7943,11 +7960,34 @@ EXPORTED void mailbox_iter_uidset(struct mailbox_iter *iter, seqset_t *seq)
     mailbox_iter_startuid(iter, seqset_first(seq));
 }
 
+EXPORTED void mailbox_iter_timer(struct mailbox_iter *iter,
+                                 struct timespec until,
+                                 unsigned nchecktime)
+{
+    iter->until = until;
+    iter->nchecktime = nchecktime;
+}
+
 EXPORTED const message_t *mailbox_iter_step(struct mailbox_iter *iter)
 {
     if (mailbox_wait_cb) mailbox_wait_cb(mailbox_wait_cb_rock);
 
     for (iter->recno++; iter->recno <= iter->num_records; iter->recno++) {
+
+        // Check timer, if any
+        if ((iter->until.tv_sec || iter->until.tv_nsec) &&
+            ((!iter->nchecktime || (iter->recno % iter->nchecktime) == 0))) {
```

Comment:
So we check every 1000 records? How guaranteed are we that that will happy very quickly? What is "very quickly"? sub second, or second, or longer?

I guess what I'm asking is if this could conceivably take 5s to process 1000 records, what use is a 2s lock timeout? And if it's not useful, should we recommend a minimum, or log if we take significantly longer than our lock timeout so user's expectations can be adjusted?

Reply:
Checking 1000 records take between 7-9 milliseconds on my consumer-grade NVME. I updated the man page to state that lock durations are logged at verbosity level 3. This should be enough to get started debugging long locks on extremely slow disks, and we can still add more knobs to lock configuration if required.

Reply:
Great